matrix-e2e.c

```diff
@@ -680,6 +680,119 @@ int matrix_e2e_get_device_keys(MatrixConnectionData *conn, const gchar *device_i
+    purple_debug_info("matrixprpl",
+                      "%s: Type %zd olm encrypted message from %s\n",
+                      __func__, (size_t)type, cevent_sender);
+    if (!type) {
+        /* A 'prekey' message to establish an Olm session
+         * TODO!!!!: Try existing sessions and check with
+         * matches_inbound_session */
+        session = olm_session(g_malloc0(olm_session_size()));
+        const gchar *cevent_body;
+        cevent_body = matrix_json_object_get_string_member(our_ciphertext,
+                                                               "body");
+        gchar *cevent_body_copy = g_strdup(cevent_body);
+        if (olm_create_inbound_session_from(session, conn->e2e->oa, sender_key,
+                                               strlen(sender_key),
+                                        cevent_body_copy, strlen(cevent_body))
+            == olm_error()) {
+            purple_debug_info("matrixprpl",
+                              "%s: prekey inbound_session_from failed : %s\n",
+                    __func__, olm_session_last_error(session));
+            goto err;
+        }
+        if (olm_remove_one_time_keys(conn->e2e->oa, session) == olm_error()) {
+            purple_debug_info("matrixprpl",
+                              "%s: Failed to remove 1tk, inbound session: %s\n",
+                    __func__, olm_account_last_error(conn->e2e->oa));
+            goto err;
+        }
+        cevent_body_copy = g_strdup(cevent_body);
+        size_t max_plaintext_len = olm_decrypt_max_plaintext_length(session,
+                                       0 /* Prekey */,
+                                       cevent_body_copy,
+                                       strlen(cevent_body_copy));
+        if (max_plaintext_len == olm_error()) {
+            purple_debug_info("matrixprpl",
+                              "%s: Failed to get plaintext length %s\n",
+                              __func__, olm_session_last_error(session));
+            goto err;
+        }
+        plaintext = g_malloc0(max_plaintext_len + 1);
+        cevent_body_copy = g_strdup(cevent_body);
+
+        size_t pt_len = olm_decrypt(session, 0 /* Prekey */, cevent_body_copy,
+                                       strlen(cevent_body),
+                                       plaintext, max_plaintext_len);
+        if (pt_len == olm_error() || pt_len >= max_plaintext_len) {
+            purple_debug_info("matrixprpl",
+                              "%s: Failed to decrypt inbound session creation"
+                              " event: %s\n",
+                              __func__, olm_session_last_error(session));
+            goto err;
+        }
+        plaintext[pt_len] = '\0';
+        // TODO: Store session in db
+    } else {
+        purple_debug_info("matrixprpl", "%s: Type %zd olm\n", __func__, type);
+    }
+    // TODO  resave account? Or just session?
+    //
+    g_free(plaintext);
```

Comment:
yeah, memset doesn't work. 

http://matrix.org/git/olm/tree/src/memory.cpp#n24 is @NegativeMjark's C++ implementation - the key is basically having `volatile` everywhere. I've looked at the disassembly of the result and satisfied myself that it looks like it does the right thing. I'd guess you could take out the C++-specific magic and end up with something that works.

Or we could expose it as a utility in olm.h.